### PR TITLE
Post fine performance metrics to Spans

### LINKED ISSUE #7776: Fine performance metrics: apportion to Computations
- Part of https://github.com/dask/distributed/issues/7665

Computation objects are commonly used by third-party Scheduler plugins (e.g. Coiled Analytics) to visualize data.
They can be crudely defined as everything that happened to the cluster between two moments when it was idle.
As there should be only one active at any given time, it makes perfect sense to apportion the metrics to the Computation objects, in addition to the Scheduler's global ever-increasing counter.

### PR DESCRIPTION
**Pull request by crusaderky**
- Part of #7860
- Closes #7776

Out of scope: make the Bokeh GUI span-aware

### CODE REVIEW

File: distributed/dashboard/components/scheduler.py [hunk 1]
```diff
@@ -3421,11 +3421,16 @@ def format(self, unit: str, val: Any) -> str:
...
-        )
+        items = defaultdict(float)
+        for k, v in self.scheduler.cumulative_worker_metrics.items():
+            if isinstance(k, tuple) and k[0] == "get-data":
+                items[k] = v
+            elif isinstance(k, tuple) and k[0] == "execute":
+                items[k[:1] + k[2:]] += v  # sum all span_id's together
+        # Note: this sort works because we removed span_id's. Otherwise, it would crash
+        # when the spans extension is disabled and span_id's are None as a consequence.
+        items = sorted(items.items())
```
Comment (crusaderky) [inline distributed/dashboard/components/scheduler.py:3432]: FYI @milesgranger 
Support in bokeh for the spans is highly desirable but I'd rather leave it to a different PR.

File: distributed/worker_state_machine.py [hunk 2]
```diff
@@ -3604,12 +3604,13 @@ def __init__(self, state: WorkerState):
...
         self._async_instructions = set()
 
-    def _start_async_instruction(
+    def _start_async_instruction(  # type: ignore[valid-type]
         self,
         task_name: str,
         func: Callable[P, Awaitable[StateMachineEvent]],
         /,
         *args: P.args,
+        span_id: str | None = None,
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:3613]: By the time the async instruction finishes, the task may have been forgotten, so we need to preserve the span_id. If we didn't, time wasted by work stealing would quietly disappear from all reports.

**Comment by crusaderky** [comment]
A test is failing; investigating
```
  File "/Users/runner/work/distributed/distributed/distributed/spans.py", line 423, in heartbeat
    _, span_id, prefix, activity, unit = k
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: too many values to unpack (expected 5)
```

File: distributed/spans.py [hunk 3]
```diff
@@ -262,8 +265,25 @@ def nbytes_total(self) -> int:
...
+    @property
+    def cumulative_worker_metrics(self) -> defaultdict[tuple[str, ...], float]:
+        """Replica of Worker.digests_total and Scheduler.cumulative_worker_metrics, but
+        only for the metrics that can be attributed to the current span tree.
+        The span_id has been removed from the key.
+
+        At the moment of writing, all keys are
+        ``("execute", <task prefix>, <activity>, <unit>)``
+        but more may be added in the future with a different format; please test for
+        ``k[0] == "execute"``.
```
Comment (milesgranger) [inline distributed/spans.py:277]: Do you think it'd be possible later on to map everything to some dataclass or similar? There's an awful lot of tedious checks building up for key lengths / values to determine what's what.

**Review by milesgranger — APPROVED** [review]
Overall seems good to me, baring the test failure you're already looking into.

File: distributed/spans.py (hunk 3 above)
Comment (crusaderky) [reply distributed/spans.py:277]: it is certainly possible and sensible, yes

**Comment by crusaderky** [comment]
This is ready for final review. @milesgranger could you re-review distributed/dashboard/components/scheduler.py? I just did major changes to it.

**Review by milesgranger — APPROVED** [review]
The overhaul w/ `get_metrics` is an improvement IMO, thanks!
